compiler/src/dotty/tools/dotc/semanticdb/SymbolOps.scala

```diff
@@ -0,0 +1,218 @@
+            case v: s.ValueSignature =>
+              val tparamss = sym.rawParamss.filter(ps => ps.forall(_.isTypeParam))
+              val stparams = tparamss.flatten.sscope
+              s.ValueSignature(s.UniversalType(Some(stparams), v.tpe))
+            case _ => s.Signature.Empty
+          }
+
+        case other =>
+          s.ValueSignature(
+            other.toSemanticType
+          )
+      }
+      loop(tpe)
+
+    private def toSemanticType(using LinkMode, Context): s.Type =
+      import ConstantOps._
+      def loop(tpe: Type): s.Type = tpe match {
+        case ExprType(tpe) =>
+          val stpe = loop(tpe)
+          s.ByNameType(stpe)
+
+        case TypeRef(pre, desig) if desig.isInstanceOf[Symbol] =>
+          val spre = if(tpe.hasTrivialPrefix) s.Type.Empty else loop(pre)
+          val ssym = symbolName(desig.asInstanceOf[Symbol])
+          s.TypeRef(spre, ssym, Seq.empty)
+
+        case TermRef(pre, desig) if desig.isInstanceOf[Symbol] =>
+          val spre = if(tpe.hasTrivialPrefix) s.Type.Empty else loop(pre)
+          val ssym = symbolName(desig.asInstanceOf[Symbol])
+          s.SingleType(spre, ssym)
+
+        // TODO: TypeParamRef and TermParamRef
+        // but how can we retrieve their symbols? (for representing them as TypeRef and TermRef on Semanticdb)
+        // or should we add new types to Semanticdb
+
+        case ThisType(TypeRef(_, desig)) if desig.isInstanceOf[Symbol] =>
+          val ssym = symbolName(desig.asInstanceOf[Symbol])
+          s.ThisType(ssym)
+
+        case SuperType(thistpe, supertpe) =>
+          val spre = loop(thistpe.typeSymbol.info)
+          val ssym = symbolName(supertpe.typeSymbol)
+          s.SuperType(spre, ssym)
+
+        // val clazzOf = classOf[...]
+        case ConstantType(const) if const.tag == core.Constants.ClazzTag =>
+          loop(const.typeValue)
+
+        case ConstantType(const) =>
+          s.ConstantType(const.toSemanticConst)
+
+        case rt @ RefinedType(parent, name, info) =>
+          // `X { def x: Int; def y: Int }`
+          //
+          // RefinedType(
+          //   parent = RefinedType(
+          //     parent = TypeRef(..., X)
+          //     ...
+          //   )
+          //   refinedName = x
+          //   refinedInfo = TypeRef(..., Int)
+          // )
+          def getParent(parent: Type): Type = parent match {
+            case RefinedType(p, _, _) => getParent(p)
+            case _ => parent
+          }
+
+          val realParent = getParent(parent)
+          val stpe = realParent match {
+            // val tp: M with N { def k: Int } = ???
+            case AndType(x, y) =>
+              s.WithType(Seq(loop(x), loop(y))) // TODO: for M with N with L
+            case _ =>
+              s.WithType(Seq(loop(parent)))
+          }
+          // TODO: collect all refinement symbols
+          // val sdecls = decls.sscope(using LinkMode.HardlinkChildren)
```

Comment:
here the symbols can be looked up by collecting the refined names and infos into a list, and then look them up with
```scala
for (name, info) <- refinedInfos yield
  val pre = ctx.typeAssigner.maybeSkolemizePrefix(rt, name)
  val denot = rt.findMember(name, pre)
  assert(denot.info eq info, s"(${denot.info.show}) is not eq to (${info.show})")
  val sym =
    if denot.symbol.exists then
      println(s"refined ${denot.symbol.show}")
    else
      ???
```

Reply:
also worth noting that `PolyFunction { def apply[T](t: T): T }` and the like will never have a symbol - what to do here?

Reply:
Thank you for the comment! Tried to collect symbols by `Types#findMember` https://github.com/lampepfl/dotty/pull/12885/commits/ebdf90014f270a1ae9fed670ca7aa5193f4c307b
I found `Types.findMember` can lookup the denotation with Type information, but it looks like those denotations always have `NoSymbol`.

For example, 

```scala
val x: { val inner: Int } = new { val inner = 1 }

// where <val x>.info is
RefinedType(
  parent = TypeRef(
    prefix = ThisType(tref = TypeRef(prefix = NoPrefix, myDesignator = module class lang)),
    myDesignator = class Object
  ),
  refinedName = inner,
  refinedInfo = TypeRef(
    prefix = TermRef(
      prefix = ThisType(tref = TypeRef(prefix = NoPrefix, myDesignator = module class <root>)),
      myDesignator = object scala
    ),
    myDesignator = class Int
  )
)

// <val x>.info.findMember(<val x>, inner)
// return denotaiton with info, but NoSymbol
```

Besides, for the polymorphic function type, as you mentioned `findMember` fails to find the symbol of the refinement `def apply`, and it instead finds `apply` method of `Function1`.

```scala
package scala3types
// with
trait Entry { type Key; val key: Key }
def extractKey(e: Entry): e.Key = e.key

// Function1[Entry, Entry#Key] { def apply(e: Entry): Entry#Key  }
val extractor: (e: Entry) => e.Key = extractKey
```

<details>

```scala
RefinedType(
  parent = AppliedType(
    tycon = TypeRef(
      prefix = ThisType(tref = TypeRef(prefix = NoPrefix, myDesignator = module class scala)),
      myDesignator = trait Function1
    ),
    args = List(
      TypeRef(
        prefix = ThisType(
          tref = TypeRef(prefix = NoPrefix, myDesignator = module class scala3types)
        ),
        myDesignator = trait Entry
      ),
      TypeRef(
        prefix = TypeRef(
          prefix = ThisType(
            tref = TypeRef(prefix = NoPrefix, myDesignator = module class scala3types)
          ),
          myDesignator = trait Entry
        ),
        myDesignator = type Key
      )
    )
  ),
  refinedName = apply,
  refinedInfo = MethodType(paramNames = List(e))
)
```

// =>



```scala
ValueSignature(
  tpe = StructuralType(
    tpe = WithType(
      types = List(
        TypeRef(
          prefix = Empty,
          symbol = "scala/Function1#",
          typeArguments = List(
            TypeRef(prefix = Empty, symbol = "scala3types/Entry#", typeArguments = List()),
            TypeRef(
              prefix = TypeRef(
                prefix = Empty,
                symbol = "scala3types/Entry#",
                typeArguments = List()
              ),
              symbol = "scala3types/Entry#Key#",
              typeArguments = List()
            )
          )
        )
      )
    ),
    declarations = Some(
      value = Scope(
        symlinks = List(),
        hardlinks = List(
          SymbolInformation(
            symbol = "scala/Function1#apply().",
            language = SCALA,
            kind = METHOD,
            properties = 0,
            displayName = "apply",
            signature = MethodSignature(
              typeParameters = Some(value = Scope(symlinks = List(), hardlinks = List())),
              parameterLists = List(
                Scope(
                  symlinks = List(),
                  hardlinks = List(
                    SymbolInformation(
                      symbol = "scala/Function1#apply().(v1)",
                      language = SCALA,
                      kind = PARAMETER,
                      properties = 0,
                      displayName = "v1",
                      signature = ValueSignature(
                        tpe = TypeRef(
                          prefix = Empty,
                          symbol = "scala/Function1#[T1]",
                          typeArguments = List()
                        )
                      ),
                      annotations = List(),
                      access = Empty,
                      overriddenSymbols = List(),
                      documentation = None
                    )
                  )
                )
              ),
              returnType = TypeRef(
                prefix = Empty,
                symbol = "scala/Function1#[R]",
                typeArguments = List()
              )
            ),
            annotations = List(),
            access = Empty,
            overriddenSymbols = List(),
            documentation = None
          )
        )
      )
    )
  )
)
```
</details>

I'm wondering, we should work around this problem by

- (option1) Create dummy symbols for Semanticdb
- (option2) Pass around the tree to `toSemanticSig` and `toSemanticType` and get refinements symbols from tree.

and I'm currently leaning to option2, because with option1 I guess we have to make up a lot of dummy information for the symbol to generate SymbolInformation from it (for [hardlinking](https://scalameta.org/docs/semanticdb/specification.html#scope)).

Reply:
BTW, at least in `ExtractSemanticDB` phase, `maybeSkolemizePrefix` seems not needed because it always returns `qualType` https://github.com/lampepfl/dotty/blob/ebdf90014f270a1ae9fed670ca7aa5193f4c307b/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala#L115 outside of `Typer` phase. (because `isLegalPrefix` returns `true` outside of `Typer` phase)
(I actually don't understand what exactly `SkolemType` for...)

Reply:
Here's rough implementation of generating dummy symbols for refinements in RefinedType https://github.com/lampepfl/dotty/pull/12885/commits/a46e4c7189c2a4eebfe985a7f651055d6432b747

Reply:
it seems in scala 2 implementation a local symbol is generated for each refinement, and in the case of a refinement where the denotation of `findMember` has a symbol then you record that the refinement is an override